### PR TITLE
Add Type-Erased FlatL2Index and IVFFlatIndex

### PR DESCRIPTION
**Pull request by lums658**
This PR wraps up the `flat_l2_index` and `ivf_flat_index` classes with a type-erased interface.  It recapitulates that aspect of PR #153.

The type-erased indexes allow for vector databases to be ingested and queried.  The API is modeled after those, e.g., in FAISS and pinecone.

On ingestion, the user can specify the desired types of the index.
```
  auto a = IndexIVFFlat(std::make_optional<IndexOptions>(
      {{"feature_type", "float32"},
       {"id_type", "uint32"},
       {"px_type", "uint32"}}));
```

Vector inputs can be specified in a type oblivious fashion:
```
  auto training_set = FeatureVectorArray(ctx, siftsmall_inputs_uri);
```

The index is created with 
```
  a.train(training_set, kmeans_init::random);
  a.add(training_set);
  a.write_index(ctx, api_ivf_flat_index_uri, true);
```

Queries can similarly be read and applied without specifying the type:
```
   auto query_set = FeatureVectorArray(ctx, siftsmall_query_uri);
   auto&& [s, t] = a.query_infinite_ram(query_set, k_nn, nprobe);
```

This functionality is exposed in Python via the pybind11 wrappers in `apis/python/src/tiledb/vector_search/type_erased_module.cc``.

An index can be built in a straightforward way with the Python bindings
```
        a = vspy.IndexIVFFlat(id_type="uint32", px_type="uint32")

        training_set = vspy.FeatureVectorArray(ctx, siftsmall_inputs_uri)

        a.train(training_set, "random");   # or kmeanspp
        a.add(training_set);
```

And read and queried
```
	a = vspy.IndexIVFFlat(ctx, siftsmall_group_uri)
	query_set = vspy.FeatureVectorArray(ctx, siftsmall_query_uri)

        s, t = a.query_finite_ram(query_set, k_nn, nprobe, 1000)
```

### CODE REVIEW

File: src/include/api/flat_l2_index.h [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+#include "api/feature_vector.h"
+#include "api/feature_vector_array.h"
+#include "api_defs.h"
+#include "detail/linalg/tdb_helpers.h"
+#include "index/flat_l2_index.h"
+
+/**
+ * A type-erased index class for C++ ivf_flat_l2_index.
+ */
+class IndexFlatL2 {
```
Comment (NikolaosPapailiou) [inline src/include/api/flat_l2_index.h:49]: Shouldn't this be just `IndexFlat`? Why do we add the distance name in the class name when we want these classes to be using configurable distance functions?

**Review by NikolaosPapailiou — APPROVED** [review]

File: apis/python/test/test_utils.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import os
 import numpy as np
+from array_paths import *
 from tiledb.vector_search.utils import load_fvecs, load_ivecs, write_fvecs, write_ivecs
 from array_paths import *
```
Comment (jparismorgan) [inline apis/python/test/test_utils.py:5]: (Minor) Can remove, looks like a duplicate of line 5.

File: src/include/api/flat_l2_index.h [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+      // @note We need to maintain same layout -> or swap extents
+      switch (dtype) {
+        case TILEDB_FLOAT32: {
+          auto qspan = MatrixView<float, stdx::layout_left>{
+              (float*)vectors.data(),
+              extents(vectors)[0],
+              extents(vectors)[1]};  // @todo ??
+          auto [s, t] = impl_index_.query(qspan, k_nn);
+          debug_slice(t);
```
Comment (jparismorgan) [inline src/include/api/flat_l2_index.h:228]: (Minor) Remove?

File: src/include/api/ivf_flat_index.h [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<uint8_t, uint64_t, uint64_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<float, uint64_t, uint64_t>>>(
+          ctx, group_uri, config);
+    }
```
Comment (jparismorgan) [inline src/include/api/ivf_flat_index.h:222]: (Optional) Might be nice to add an else case and throw an exception. That way if we add to the enum and not here we can catch it more easily. Same with the other big if/else if below.

File: src/include/api/ivf_flat_index.h [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+#if 0
+  constexpr size_t ntotal() const {
+    // @todo
+    return 0;
+  }
+
+  constexpr auto num_vectors() const {
+    return _cpo::num_vectors(*index_);
+  }
+#endif
```
Comment (jparismorgan) [inline src/include/api/ivf_flat_index.h:403]: (Minor) Remove?

File: src/include/api/ivf_flat_index.h [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+    [[nodiscard]] virtual size_t dimension() const = 0;
+
+    [[nodiscard]] virtual size_t num_partitions() const = 0;
+
+// Don't think we need these
+#if 0
+    virtual size_t ntotal() const = 0;
+
+    virtual size_t num_vectors() const = 0;
+#endif
```
Comment (jparismorgan) [inline src/include/api/ivf_flat_index.h:478]: (Minor) Remove?

File: src/include/api/flat_l2_index.h [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+      auto dtype = vectors.feature_type();
+
+      // @note We need to maintain same layout -> or swap extents
+      switch (dtype) {
+        case TILEDB_FLOAT32: {
+          auto qspan = MatrixView<float, stdx::layout_left>{
+              (float*)vectors.data(),
+              extents(vectors)[0],
+              extents(vectors)[1]};  // @todo ??
```
Comment (jparismorgan) [inline src/include/api/flat_l2_index.h:226]: There are quite a few of these `// @todo ??` comments - would suggest to add comments about what the todo is saying, i.e. `// @todo Update to __`, as it's not clear just from reading the code.

File: src/include/api/ivf_flat_index.h [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+    // Don't think we need these
+#if 0
+    size_t ntotal() const override {
+      return ::num_vectors(impl_index_);
+    }
+
+    size_t num_vectors() const override {
+      return ::num_vectors(impl_index_);
+    }
+#endif
```
Comment (jparismorgan) [inline src/include/api/ivf_flat_index.h:687]: (Minor) Remove?

File: src/include/detail/flat/qv.h [hunk 9]
```diff
@@ -134,15 +141,22 @@ auto qv_query_heap(
...
     T,
     const DB& db,
     const Q& query,
     const ID& ids,
     int k_nn,
-    unsigned nthreads) {
+    unsigned nthreads,
+    Distance distance = Distance{}) {
   scoped_timer _{tdb_func__};
+  // load(db);
```
Comment (jparismorgan) [inline src/include/detail/flat/qv.h:159]: (Minor) What's going on here? Should we load? If not, let's remove this?

File: src/include/detail/flat/qv.h [hunk 10]
```diff
@@ -182,9 +196,46 @@ auto qv_query_heap(
...
+    unsigned nthreads,
+    Distance distance = Distance{}) {
+  return qv_query_heap(
+      without_ids{},
+      db,
+      q,
+      std::vector<uint64_t>{},
+      k_nn,
+      nthreads,
+      distance);  /// ????
```
Comment (jparismorgan) [inline src/include/detail/flat/qv.h:216]: (Minor) What's going on here?

File: src/include/detail/flat/qv.h [hunk 11]
```diff
@@ -357,18 +393,46 @@ auto qv_query_heap_tiled(
   return top_k;
 }
 
+template <feature_vector_array DB, feature_vector_array Q>
+auto qv_query_heap_tiled(DB& db, const Q& q, int k_nn, unsigned nthreads) {
+  return qv_query_heap_tiled(
+      without_ids{}, db, q, std::vector<uint64_t>{}, k_nn, nthreads);  // ????
```
Comment (jparismorgan) [inline src/include/detail/flat/qv.h:399]: (Minor) What's going on here?

File: src/include/index/index_defs.h [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+         {"input_vectors_array_name", "input_vectors"},
+         {"external_ids_array_name", "external_ids"},
+         {"updates_array_name", "updates"},
+
+         // Not arrays
+         // @todo Where are these stored?
+         {"default_attr_filters", "zstd"},
+         {"support_timetravel", "true"},
+     }},
+};
```
Comment (jparismorgan) [inline src/include/index/index_defs.h:134]: (Unsure) Why does this one not have `{"partial_write_array_dir", "temp_data"}`? Will this cause a crash if we try to access it? In `apis/python/src/tiledb/vector_search/storage_formats.py`, the 0.3 version still has it.

File: src/include/index/ivf_flat_metadata.h [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+   * @param check
+   * @return
+   */
+  auto check_string_metadata(
+      tiledb::Group& read_group, const metadata_string_check_type& check) {
+    tiledb_datatype_t v_type;
+    uint32_t v_num;
+    const void* v;
+
+    auto&& [name, value, required] = check;  // copilot filled in "required"
```
Comment (jparismorgan) [inline src/include/index/ivf_flat_metadata.h:178]: (Minor) Remove comment, and also below.

**Review by jparismorgan — APPROVED** [review]
Mostly minor cleanups, generally looks good to me.

File: src/include/api/flat_l2_index.h (hunk 1 above)
Comment (lums658) [reply src/include/api/flat_l2_index.h:49]: The C++ `ivf_index` class should be parameterized (will be parameterized) by distance function, but the type-erased class can't be type-parameterized in the same way a class template can.  I think Faiss has `indexFlatL2`, for instance (https://faiss.ai/cpp_api/struct/structfaiss_1_1IndexFlatL2.html). 

I created a story for this: https://app.shortcut.com/tiledb-inc/story/40228/parameterize-c-index-classes-and-their-type-erased-wrappers-on-distance-function

File: src/include/api/ivf_flat_index.h (hunk 8 above)
Comment (lums658) [reply src/include/api/ivf_flat_index.h:687]: Removed `ntotal()` and `num_vectors` from both type-erased index classes.

File: src/include/detail/flat/qv.h (hunk 9 above)
Comment (lums658) [reply src/include/detail/flat/qv.h:159]: This touches on an issue we need to clean up:
https://app.shortcut.com/tiledb-inc/story/40229/specify-definitive-behavior-for-out-of-core-algorithms-and-data-structures

File: src/include/index/index_defs.h (hunk 12 above)
Comment (lums658) [reply src/include/index/index_defs.h:134]: This is fine.  "partial_write_array_name" is just a key for "temp_data".  I just gave all the keys consistent names (x_y_array_name).

**Pull request merged**